lecture_5/homework/app/services/reports/siege_report.rb

```diff
@@ -3,10 +3,23 @@
 module Reports
   class SiegeReport
     def initialize(building:)
+      @building = building
+      @warriors = building.warriors
     end
 
     def call
-      raise NotImprementedYet
+      @building.siege_ability = (@warriors.present? ? @building.granary / food_demand : 0)
+      @building.save
```

Comment:
To zapisywanie atrybutu niepotrzebnie dociąża nam ten serwis, powinno też się znaleźć "wyżej" :) Wiąże się to też między innymi ze wzorcem dobrego projektowania aplikacji, Single Responsibility Principle.  Ogólnie możnaby zoptymalizować nasz callback przez obsługę wyliczenia tych dni i zapisania tej wartości na budynku w kolejce asynchronicznej np. Sidekiq, Resque, ActiveJob :D Albo przez w ogóle zastanowieniem się czy nie wystarczy liczyć tych dni w serializerze, który by potrafił scachować wywołanie tego serwisu na jakiś czas, żeby oszczędzić pracy apce, ale to są tematy których nie było na LevelUp, nie mniej są bardzo ważne i polecam je zbadać :) napewno pomyślimy o nich w kolejnych edycjach, a napewno będą na praktykach :D